man/bppreroot.1
.\" SPDX-FileCopyrightText: The Bio++ Development Group
.\"
.\" SPDX-License-Identifier: CECILL-2.1

.TH BPPREROOT 1 LOCAL

.SH NAME

bppreroot - Sequential rerooting of trees with Bio++

.SH SYNOPSIS

.B bppreroot [options]

.SH AVAILABILITY

All UNIX flavors

.SH DESCRIPTION

bppreroot reroots a set of trees according to a list of outgroup species.

.SH OPTIONS

You should refer to 'info bppsuite' or to the online manual of bppsuite for a complete list of available options.

.TP 5

--noninteractive

generates output for redirection in a file.

.TP

param=file

reads a file for loading options

.SH AUTHOR

The Bio++ Development Team.


ChangeLog
19/03/18 -*- Version 2.4.0 -*-

19/02/18 Julien Dutheil
* Bug fixed in bppPopStats: dN and dS were inverted!

10/12/17 -*- Version 2.3.2 -*-

06/11/17 Julien Dutheil
* Added estimation of kappa to bppPopStats + bugs fixed

06/06/17 -*- Version 2.3.1 -*- 

10/05/17 -*- Version 2.3.0 -*- 

21/12/16 Julien Dutheil
* BppPhySamp migrated to its own package.

17/12/16 Julien Dutheil
* Improved bppPopStats, outgroup analyses, MK table and output to file.

06/10/16 Julien Dutheil
* bppSeqGen now takes Hudson's MS ARG files as input.
* new input.tree.scale option to adjust branch lengths.

22/05/16 Julien Dutheil
* bpppopstats now computes codon statistics per site (e.g. PiN, PiS)
* Added example for bpppopstats.

29/06/15 Julien Dutheil
* New program bppPopStats

28/01/15 Julien Dutheil
* PhySamp now outputs the correct tree!

25/11/14 Julien Dutheil
* bppalnscore output contiguous segments in mase selection.
* bppalnscore site selection works with codon alphabets.

24/11/14 Julien Dutheil
* bppSeqMan can now read alignments and perform site selection priori to any analysis.

28/09/14 -*- Version 2.2.0 -*- 

16/09/14 Julien Dutheil
* bppSeqGen supports generic characters as ancestral states in info file (bug #87).

12/08/14 Julien Dutheil
* bppSeqGen supports simulation under Markov-Modulated models.

06/06/14 Julien Dutheil
* bppPhySamp now output trees as well.

17/04/14 Julien Dutheil
* bppSeqMan: Translate uses global genetic_code argument instead of local 'code' argument.
* All programs use the --seed argument to set random seed for reproduceable results.

08/03/13 -*- Version 0.8.0 -*-

22/01/13 Julien Dutheil
* Compiles with -Wconversion

14/01/13 Mathieu Groussin
* Added COaLA model for proteins in bppML and bppSeqGen.

23/11/12 Laurent Guéguen
* New bppMixedLikelihood program

08/02/12 -*- Version 0.7.0 -*-

21/12/11 Julien Dutheil
* New bppAlnScore program
* Improved bppSeqMan: protein translation now works + more stop codons removal options.

17/06/11 Julien Dutheil
* Improved bppancestor, also allows to compute frequencies for leaves.

11/06/11 -*- Version 0.6.2 -*-
* Small bug fixed + new packages.

03/06/10 -*- Version 0.6.0 -*-
* Compatibility with Bio++ 2.0.0.

03/06/10 Julien Dutheil
* New method in bppPhySamp + new, faster, implementation.

25/03/10 -*- Version 0.5.0 -*-

22/08/09 Julien Dutheil
* Added CMake configuration files, with twick for building the manual (assumes
		texinfo is installed).

22/07/09 Julien Dutheil
* Added program bppTreeDraw.

30/06/09 Julien Dutheil
* BppAncestor: now output expected ancestral states frequencies.
* BppAncestor: new method to compute EB estimates of ancestral states frequencies.

10/06/09 -*- Version 0.4.0 -*-

22/05/09 Julien Dutheil
* Keyval syntax for sequence format description.
* Documentation and example update.
* SeqMan now uses the SequenceApplicationTools method.

15/05/09 Julien Dutheil
* Compatibility update with the libs, new keyval syntax for model
specification.
* Documentation up to date.

02/02/09 Julien Dutheil
* bppConsense now outputs the number of leaves of the input tree.
* New progress bar in bppConsense.

08/01/09 Julien Dutheil
* Added texinfo documentation.
* More bootstrap options in BppDist (the same as in BppML).

25/09/08 -*- Version 0.3.1 -*-

07/12/08 Bastien Boussau & Julien Dutheil
* Bug removed in non-homogeneous models with covarions.

09/10/08 Julien Dutheil
* New action 'invert' in bppSeqMan.

25/09/08 -*- Version 0.3.0 -*-

20/09/08 Julien Dutheil
* New program bppAncestor to reconstruct acestral states and rates.
* New +F models from PhylLib. 

02/07/08 Laurent Guegen
* Bug fixed in bppML: frequencies can be computed from the data with
non-homogeneous models.

19/06/08 Julien Dutheil
* Shell variables are now 'global' in option files.

24/04/08 -*- Version 0.2.3 -*-
* This version requires the CVS version of Bio++ to be built.

03/03/08 Julien Dutheil
* Bug fixed in keep_complete method of BppSeqMan.

27/03/08 Julien Dutheil
* Bug fixed when printing parameters of non-homogeneous models in bppML.

22/01/08 -*- Version 0.2.2 -*-

28/02/08 Celine Scronavacca
* Bug fixed in ReRoot.

22/01/08 -*- Version 0.2.1 -*-

06/02/08 Celine Scornavacca
* Bug fixed in ReRoot + correction of version number in bppSeqGen.

22/01/08 -*- Version 0.2.0 -*-

13/01/08 Julien Dutheil
* More options in bppml: possibility to chose between simple and double
recursion and data compression method.
* Simple recursion method improved.

11/01/08 Julien Dutheil
* Compatibility update
* bppML now output more handy parameter values.

31/12/07 Julien Dutheil
* New program Phylo Sampler.

07/11/07 Julien Dutheil
* Compatibility update with Bio++ cvs.
* Bug fixed in BppSeqGen.
* BppSeqGen can now simulate data from a set of trees (ancestral
  recombination graph).

25/10/07 Julien Dutheil
* Several bug fixed in BppML and BppSeqGen with non-homogeneous models.
* Examples updated + new example option files for non-homogeneous models.

05/10/07 Bastien Boussau & Julien Dutheil
* Non-homogenous sequence simulation in BppSeqGen.
* Model options in files have changed to suit the non-homogeneous case.
  The covarion option now is fully part of the model options, following the
	new syntax 'model=modelname+covarionname', eg 'T92+G2001'.

03/10/07 Julien Dutheil
* New program Bio++ Sequence Manipulator

21/09/07 Julien Dutheil
* Bug fixed in BppDist: branch length are indeed not estimated when parameters
  to ignore contains "BrLen".
* Stop condition when using iterations is now based on topological distance
  instead of likelihood comparisons (phyllib).
* New option optimization.topology_nni.method = [fast|better|phyml]

17/06/07 -*- Version 0.1.0 -*-

16/07/06 Julien Dutheil
* New option bootstrap.verbose for bppML.
* Bug fixed in bootstrap computation.

06/06/06 Julien Dutheil
* New program bppConsense.

05/05/07 Julien Dutheil
* First alpha release of bppsuite.

-*- Old bppML ChangeLog -*-

23/04/07 Julien Dutheil

bppSuite/bppReRoot.cpp
// SPDX-FileCopyrightText: The Bio++ Development Group
//
// SPDX-License-Identifier: CECILL-2.1

// From the STL:
#include <iostream>

using namespace std;

// From bpp-core:
#include <Bpp/Version.h>
#include <Bpp/Numeric/VectorTools.h>
#include <Bpp/Text/TextTools.h>
#include <Bpp/Text/StringTokenizer.h>
#include <Bpp/App/BppApplication.h>
#include <Bpp/App/ApplicationTools.h>

// From PhylLib:
#include <Bpp/Phyl/Tree/Tree.h>
#include <Bpp/Phyl/Tree/Node.h>
#include <Bpp/Phyl/Tree/TreeExceptions.h>
#include <Bpp/Phyl/Tree/TreeTemplateTools.h>
// From bpp-phyl:
#include <Bpp/Phyl/Io/Newick.h>
#include <Bpp/Phyl/App/PhylogeneticsApplicationTools.h>

using namespace bpp;

typedef TreeTemplate<Node> MyTree;


void help()
{
  (*ApplicationTools::message << "bppreroot parameter1_name=parameter1_value").endLine();
  (*ApplicationTools::message << "      parameter2_name=parameter2_value ... param=option_file").endLine();
  (*ApplicationTools::message).endLine();
  (*ApplicationTools::message << "  Refer to the Bio++ Program Suite Manual for a list of available options.").endLine();
}


int main(int args, char** argv)
{
  cout << "******************************************************************" << endl;
  cout << "*                  Bio++ ReRoot, version " << BPP_VERSION << "                   *" << endl;
  cout << "* Author: C. Scornavacca                    Created     15/01/08 *" << endl;
  cout << "*                                           Last Modif. " << BPP_REL_DATE << " *" << endl;
  cout << "******************************************************************" << endl;
  cout << endl;

  if (args == 1)
  {
    help();
    return 0;
  }

  try
  {
    BppApplication bppreroot(args, argv, "BppReRoot");
    bppreroot.startTimer();

    Newick newick;
    string listPath = ApplicationTools::getAFilePath("input.list.file", bppreroot.getParams());
    ApplicationTools::displayResult("Input list file", listPath);
    if (listPath == "none")
      throw Exception("You must provide an input tree list file.");

    string outgroupsPath = ApplicationTools::getAFilePath("outgroups.file", bppreroot.getParams());
    ApplicationTools::displayResult("Outgroups file", outgroupsPath);
    if (outgroupsPath == "none")
      throw Exception("You must provide an outgroup list file.");

    string outputPath = ApplicationTools::getAFilePath("output.trees.file", bppreroot.getParams(), true, false);
    ApplicationTools::displayResult("Output file", outputPath);
    if (outputPath == "none")
      throw Exception("You must provide an output file.");

    bool printOption = ApplicationTools::getBooleanParameter("print.option", bppreroot.getParams(), false);
    bool tryAgain = ApplicationTools::getBooleanParameter("tryAgain.option", bppreroot.getParams(), true);

    vector<Tree*> tempTrees;
    vector<MyTree*> trees;

    // ApplicationTools::displayResult("Number of trees found", TextTools::toString(trees.size()));

    const string path = outgroupsPath;
    ifstream file(path.c_str(), ios::in);
    string temp, description, taxon;

    vector< vector<string>> levelOutgroup;

    // Reading outgroup levels
    while (!file.eof())
    {
      vector<string> tempTaxa;
      getline(file, temp, '\n');
      StringTokenizer line = StringTokenizer(temp, " ,");
      while (line.hasMoreToken())
      {
        tempTaxa.push_back(line.nextToken());
      }
      levelOutgroup.push_back(tempTaxa);
    }
    file.close();

    const string path2 = listPath;
    ifstream treePath(path2.c_str(), ios::in);

    if (!treePath)
    {
      throw IOException ("Newick::read: failed to read from stream");
    }

    string temp2, description2; // Initialization
    string::size_type index;

    int k = 0;

    while (!treePath.eof())
    {
      k++;
      bool printOrNot = true;
      unique_ptr<Tree> tempTree;

      getline(treePath, temp2, '\n');  // Copy current line in temporary string

      index = temp2.find(";");
      if (temp2 != "")
      {
        if (index != string::npos)
        {
          description2 += temp2.substr(0, index + 1);
          tempTree = TreeTemplateTools::parenthesisToTree(description2);
          description2 = temp2.substr(index + 1);
        }
        else
          description2 += temp;

        MyTree* tree = dynamic_cast<MyTree* >(tempTree.get());
        // ApplicationTools::displayGauge(tr, trees.size() - 1, '=');

        vector<string> leavesTree;
        leavesTree = (*tree).getLeavesNames();

        size_t numNodes = tree->getNumberOfNodes() - 1;
        size_t numNodeWithBranchLength = 0;
        vector<Node*>  nodes = tree->getNodes();
        for (size_t i = 0; i < nodes.size(); i++)
        {
          if (nodes[i]->hasDistanceToFather())
            numNodeWithBranchLength++;
        }
        if ((numNodes != numNodeWithBranchLength) && (numNodeWithBranchLength != 0)) \
        {
          cout << "Could not execute due to a source tree with missing branch lengths \n(reminder: a source tree must either have no branch length, either length for all branches\n";
          exit(-1);
        }
        vector<string> outGroup;
        bool found = false;
        bool analyseOutgroupLevel = true;
        for (size_t t = 0; t < levelOutgroup.size() && analyseOutgroupLevel; t++)
        {
          outGroup.clear();
          vector<string>::iterator Iterator;
          for (Iterator = levelOutgroup[t].begin(); Iterator != levelOutgroup[t].end(); Iterator++ )
          {
            if (VectorTools::contains(leavesTree, *Iterator))
            {
              outGroup.push_back(*Iterator);
            }
          }
          if (outGroup.size() > 0)
          {
            vector<string> remainingTaxa;
            VectorTools::diff(leavesTree, outGroup, remainingTaxa);
            if (remainingTaxa.size() > 0)
            {
              tree->newOutGroup(tree->getNode(remainingTaxa[0]));
              Node* newRoot = tree->getNode(outGroup[0]);
              vector<string>  tempLeaves = TreeTemplateTools::getLeavesNames(*newRoot);

              while (newRoot->hasFather() && !(VectorTools::containsAll(tempLeaves, outGroup)))
              {
                newRoot = newRoot->getFather();
                tempLeaves = TreeTemplateTools::getLeavesNames(*newRoot);
              }

              tempLeaves = TreeTemplateTools::getLeavesNames(*newRoot);
              std::sort(tempLeaves.begin(), tempLeaves.end());

              if (tempLeaves.size() == outGroup.size())
              {
                tree->newOutGroup(newRoot);
                found = true;
                analyseOutgroupLevel = false;
              }
              else
              {
                bool monophylOk = true;

                for (size_t f = 0; f < newRoot->getNumberOfSons() && monophylOk; f++)
                {
                  tempLeaves = TreeTemplateTools::getLeavesNames(*newRoot->getSon(f));
                  vector<string> diff;
                  VectorTools::diff(outGroup, tempLeaves, diff);

                  size_t difference = diff.size();
                  if (!( (difference == 0) || (difference == tempLeaves.size()) ) )
                  {
                    // The proposed outgroup is not monophyletic. The analysis for this tree is interrupted
                    // No more outgroup are analysed
                    monophylOk = false;
                  }
                }
                if (monophylOk)
                {
                  tempLeaves = TreeTemplateTools::getLeavesNames(*newRoot);

                  std::sort(tempLeaves.begin(), tempLeaves.end());
                  if (tempLeaves.size() != leavesTree.size())
                  {
                    MyTree* low = new MyTree(TreeTemplateTools::cloneSubtree<Node>(*newRoot));
                    tree->newOutGroup(newRoot);
                    Node* sonUpper;
                    vector<string>  tempLeaves2 = TreeTemplateTools::getLeavesNames(*(tree->getRootNode())->getSon(0));
                    std::sort(tempLeaves2.begin(), tempLeaves2.end());
                    if ((VectorTools::vectorIntersection(tempLeaves2, outGroup).size()) != 0)
                    {
                      sonUpper = (tree->getRootNode())->getSon(1);
                    }
                    else
                    {
                      sonUpper = (tree->getRootNode())->getSon(0);
                    }
                    int ident = TreeTools::getMaxId(*low, low->getRootId());
                    vector<Node*> nodesTemp = TreeTemplateTools::getNodes( *sonUpper);
                    for (size_t F = 0; F < nodesTemp.size(); F++)
                    {
                      nodesTemp[F]->setId(ident + static_cast<int>(F + 1));
                    }
                    low->getRootNode()->addSon(sonUpper);
                    tree = low;
                  }
                  // A good outgroup was found

                  found = true;
                  analyseOutgroupLevel = false;
                }
              }
            }
            if (!tryAgain)
              analyseOutgroupLevel = false;
          }
        }
        if (!found)
        {
          if (!printOption)
            printOrNot = false;
          else
            printOrNot = true;
          cout << "Sorry but I can't root your tree " << k << " ; or none of the taxa in your list is present in the tree or the outgroup is not monophyletic!\n";
        }
        else
        {
          printOrNot = (true);
          tree->resetNodesId();
        }
        if (printOrNot)
        {
          if (k == 1)
            newick.writeTree(*tree, outputPath, true);
          else
            newick.writeTree(*tree, outputPath, false);
        }

        delete tree;
      }
    }
    ApplicationTools::displayTaskDone();

    // Write rooted trees:
    for (size_t i = 0; i < trees.size(); i++)
    {
      delete trees[i];
    }

    bppreroot.done();
  }
  catch (exception& e)
  {
    cout << e.what() << endl;
    return 1;
  }

  return 0;
}
